### PR TITLE
cephfs: Implement getVolumeRootPathCeph with go-ceph

### PR DESCRIPTION
**Pull request by Madhu-1**
Implemented getVolumeRootPathCeph with go-ceph.

fixes https://github.com/ceph/ceph-csi/issues/1734

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@nixpanic is vault not added to cache? we are facing rate limit issue for vault image

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
> @nixpanic is vault not added to cache? we are facing rate limit issue for vault image

It should have been cached. #1729 might have fixed it, so I just rebased this PR to get those changes in.

**Review by nixpanic — DISMISSED** [review]
I also like to see this function become a method of the volumeOptions type.

**Comment by Madhu-1** [comment]
https://github.com/ceph/ceph-csi/issues/1734 is created to track the error handling

File: internal/cephfs/volume.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ func (vo *volumeOptions) getSubVolumeInfo(ctx context.Context, volID volumeID) (
 		if errors.Is(err, rados.ErrNotFound) {
 			return nil, ErrVolumeNotFound
 		}
-		// In case the error is other than invalid command return error to the caller.
+		// Incase the error is other than invalid command return error to the caller.
```
Comment (nixpanic) [inline internal/cephfs/volume.go]: Should this not be two words? I thought you added similar changes for this with with the new spellcheck GitHub CI job...

File: internal/cephfs/volume.go [hunk 2]
```diff
@@ -194,8 +194,9 @@ func (vo *volumeOptions) resizeVolume(ctx context.Context, volID volumeID, bytes
 			clusterAdditionalInfo[vo.ClusterID].resizeSupported = true
 			return nil
 		}
+		var invalid *fsAdmin.NotImplementedError
 		// In case the error is other than invalid command return error to the caller.
-		if !strings.Contains(err.Error(), invalidCommand) {
+		if !errors.As(err, invalid) {
```
Comment (nixpanic) [inline internal/cephfs/volume.go:200]: I think this is correct, but the more common writing seems to be
```golang
var invalid fsAdmin.NotImplementedError
if !errors.As(err, &invalid) {
    ...
```

File: internal/cephfs/volume.go [hunk 3]
```diff
@@ -88,8 +88,9 @@ func (vo *volumeOptions) getSubVolumeInfo(ctx context.Context, volID volumeID) (
 		if errors.Is(err, rados.ErrNotFound) {
 			return nil, ErrVolumeNotFound
 		}
-		// Incase the error is other than invalid command return error to the caller.
-		if !strings.Contains(err.Error(), invalidCommand) {
+		// In case the error is invalid command return error to the caller.
+		var invalid *fsAdmin.NotImplementedError
+		if errors.As(err, invalid) {
```
Comment (nixpanic) [inline internal/cephfs/volume.go:93]: same here, looks ok, but using `errors.As(err, &invalid)` seems more common

**Review by nixpanic — COMMENTED** [review]
just some nits to consider, looks good otherwise

File: internal/cephfs/volume.go (hunk 1 above)
Comment (Madhu-1) [reply internal/cephfs/volume.go]: Yes got messed up. corrected now

**Comment by Madhu-1** [comment]
@mergifyio rebase

File: internal/cephfs/volume.go [hunk 4]
```diff
@@ -194,8 +194,9 @@ func (vo *volumeOptions) resizeVolume(ctx context.Context, volID volumeID, bytes
 			clusterAdditionalInfo[vo.ClusterID].resizeSupported = true
 			return nil
 		}
+		var invalid *fsAdmin.NotImplementedError
 		// In case the error is other than invalid command return error to the caller.
-		if !strings.Contains(err.Error(), invalidCommand) {
+		if !errors.As(err, &invalid) {
```
Comment (nixpanic) [inline internal/cephfs/volume.go:200]: `invalid` is a pointer to a `fsAdmin.NotImplementedError`, it should be just `var invalid fsAdmin.NotImplementedError` instead.

File: internal/cephfs/volume.go [hunk 5]
```diff
@@ -88,8 +88,9 @@ func (vo *volumeOptions) getSubVolumeInfo(ctx context.Context, volID volumeID) (
 		if errors.Is(err, rados.ErrNotFound) {
 			return nil, ErrVolumeNotFound
 		}
-		// In case the error is other than invalid command return error to the caller.
-		if !strings.Contains(err.Error(), invalidCommand) {
+		// In case the error is invalid command return error to the caller.
+		var invalid *fsAdmin.NotImplementedError
+		if errors.As(err, &invalid) {
```
Comment (nixpanic) [inline internal/cephfs/volume.go:93]: same here, `&invalid` is now a poiner-to-a-pointer-to-a-fsAdmin.NotImplementedError

Comment (Madhu-1) [reply internal/cephfs/volume.go:93]: reference is taken from https://golang.org/pkg/errors/#example_As hope this is fine.

Comment (nixpanic) [reply internal/cephfs/volume.go:93]: I am not sure, existing usage in the current code (including `vendor/`) does not use a double pointer... This should be verified with a small test program.

Comment (Madhu-1) [reply internal/cephfs/volume.go:93]: remove double-pointer and updated the error check

**Comment by Madhu-1** [comment]
@nixpanic PTAL

**Review by nixpanic — APPROVED** [review]

**Pull request merged**